### PR TITLE
Create admin db if it doesn't exist

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR
Tested by having a static value for POSTGRES_DB in templates/postgres/deployment.yaml
And changing the values.yaml file to connect to flyteadmindb 

Without the change migrations fails
```
kubectl logs -f -n flyte flyteadmin-5f585b47b7-vch29  run-migrations
time="2022-03-09T06:51:09Z" level=info msg="Using config file: [/etc/flyte/config/cluster_resources.yaml /etc/flyte/config/db.yaml /etc/flyte/config/domain.yaml /etc/flyte/config/logger.yaml /etc/flyte/config/remoteData.yaml /etc/flyte/config/server.yaml /etc/flyte/config/storage.yaml /etc/flyte/config/task_resource_defaults.yaml]"
{"json":{"src":"migrate.go:33"},"level":"fatal","msg":"failed to connect to `host=postgres user=postgres database=flyteadmindb`: server error (FATAL: database \"flyteadmindb\" does not exist (SQLSTATE 3D000))","ts":"2022-03-09T06:51:09Z"}

2022/03/09 06:51:09 /go/src/github.com/flyteorg/flyteadmin/pkg/repositories/database.go:102
[error] failed to initialize database, got error failed to connect to `host=postgres user=postgres database=flyteadmindb`: server error (FATAL: database "flyteadmindb" does not exist (SQLSTATE 3D000))

```

With the change it fixes the issue
```
{"json":{"src":"migrate.go:55"},"level":"info","msg":"Migration ran successfully","ts":"2022-03-09T10:35:41Z"}

2022/03/09 10:35:41 /go/pkg/mod/github.com/go-gormigrate/gormigrate/v2@v2.0.0/gormigrate.go:443
[1.218ms] [rows:1] INSERT INTO migrations (id) VALUES ('2022-01-11-execution-state')
```

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue

fixes https://github.com/flyteorg/flyte/issues/1882

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/repositories/database.go [hunk 1]
```diff
@@ -94,14 +104,16 @@ func GetDB(ctx context.Context, dbConfig *runtimeInterfaces.DbConfig, logConfig
...
 			Debug:        dbConfig.DeprecatedDebug,
 		}
-		dialector = postgres.Open(getPostgresDsn(ctx, pgConfig))
+
 	default:
 		panic(fmt.Sprintf("Unrecognized database config %v", dbConfig))
 	}
-	gormDb, err := gorm.Open(dialector, &gorm.Config{
+
+	gormConfig := &gorm.Config{
```
Comment (katrogan) [inline pkg/repositories/database.go:112]: can we please initialize the dialector in the switch statements? (the whole purpose of the previous refactor so we could easily add new gorm-supported database types)

File: pkg/repositories/database.go [hunk 2]
```diff
@@ -94,14 +104,16 @@ func GetDB(ctx context.Context, dbConfig *runtimeInterfaces.DbConfig, logConfig
...
 		Logger:                                   gormLogger.Default.LogMode(logLevel),
 		DisableForeignKeyConstraintWhenMigrating: !dbConfig.EnableForeignKeyConstraintWhenMigrating,
-	})
+	}
+	gormDb, err := createIfNotExists(ctx, pgConfig, gormConfig)
```
Comment (katrogan) [inline pkg/repositories/database.go:116]: please also move this to postgres-specific case statements

File: pkg/repositories/database.go [hunk 3]
```diff
@@ -70,21 +77,24 @@ func getPostgresDsn(ctx context.Context, pgConfig runtimeInterfaces.PostgresConf
...
+// TODO : support default db names across different providers.
 func GetDB(ctx context.Context, dbConfig *runtimeInterfaces.DbConfig, logConfig *logger.Config) (
 	*gorm.DB, error) {
 	if dbConfig == nil {
 		panic("Cannot initialize database repository from empty db config")
 	}
-	var dialector gorm.Dialector
 	logLevel := getGormLogLevel(ctx, logConfig)
 
+	var pgConfig runtimeInterfaces.PostgresConfig
```
Comment (katrogan) [inline pkg/repositories/database.go:90]: no postgres specific code outside the switch statement please!

File: pkg/repositories/database.go (hunk 2 above)
Comment (pmahindrakar-oss) [reply pkg/repositories/database.go]: Done.

File: pkg/repositories/database.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply pkg/repositories/database.go:124]: makes sense. Have fixed this.

Wanted o fix that part but i think woud be good if we can do switch based on some provider attribute in the main dbConfig that way each switch case is working on deprecated, postgres, mysql ... other providers.

Let me know what you think

File: pkg/repositories/database.go (hunk 3 above)
Comment (pmahindrakar-oss) [reply pkg/repositories/database.go]: Agree. removed it.

**Comment by katrogan** [comment]
@pmahindrakar-oss what do you think of this proposal https://github.com/flyteorg/flyteadmin/pull/368 to simplify & centralize the pg-specific stuff

**Comment by pmahindrakar-oss** [comment]
Looks good @katrogan . Merging it. the reason i had kept one function was trying to make the createIfNotExists generic. But we can do that later.

**Comment by katrogan** [comment]
ah I see @pmahindrakar-oss was thrown off since of course we only have postgres impls for everything

**Review by katrogan — APPROVED** [review]

**Review by katrogan — APPROVED** [review]
